deepdiff/diff.py

```diff
@@ -558,6 +561,53 @@ def _diff_iterable(self, level, parents_ids=frozenset(), _original_type=None):
         else:
             self._diff_iterable_in_order(level, parents_ids, _original_type=_original_type)
 
+    def _compare_in_order(self, level):
+        """
+        Default compare if `iterable_compare_func` is not provided.
+        This will compare in sequence order.
+        """
+
+        return [((i, i), (x, y)) for i, (x, y) in enumerate(
+            zip_longest(
+                level.t1, level.t2, fillvalue=ListItemRemovedOrAdded))]
+
+    def _get_matching_pairs(self, level):
+        """
+        Given a level get matching pairs. This returns list of two tuples in the form:
+        [
+          (t1 index, t2 index), (t1 item, t2 item)
+        ]
+
+        This will compare using the passed in `iterable_compare_func` if available.
+        Default it to compare in order
+        """
+
+        if(self.iterable_compare_func is None):
+            # Match in order if there is no compare function provided
+            return self._compare_in_order(level)
+        try:
+            matches = []
+            y_matched = set()
+            for i, x in enumerate(level.t1):
+                x_found = False
+                for j, y in enumerate(level.t2):
+
+                    if(self.iterable_compare_func(x, y)):
+                        y_matched.add(id(y))
```

Comment:
Using the ID is cool but there could be issues when 2 things have different IDs but they are equal. Hence we usually hash the objects using DeepHash and use the hash instead. The goal is that even if the IDs are different, the hashes of 2 equal objects should be the same. I used to be concerned about the hashing being a bottleneck but so far I have not seen any bottlenecks from DeepHash application.

Reply:
I have updated to use hashes, can you please take a look and make sure I am hashing these objects correctly?